### PR TITLE
Make core available as standalone consumable modules

### LINKED ISSUE #41: Publish core conversion code as npm package
**Is your feature request related to a problem? Please describe.**

As a developer, I want to be able to fully automate my import pipeline, so that I don’t need to manually drag and drop too much.

**Describe the solution you'd like**

If the core conversion code were published as a standalone package, I would be able to easily use it directly in a custom workflow.

For instance, on my macOS machine I can attach a script to the `Downloads` folder that automatically invokes the conversion for csv files that match the expected filenames and do some further processing such as giving the files account names that correspond to the names I use in YNAB.

**Additional context**

Thanks for your work and releasing this! 🙏

### PR DESCRIPTION
**Pull request by alloy**
Opening this early and unpolished to illustrate what it could look like and to discuss requirements. I tried to stay as close as possible to your original intent/style, but can totally understand if this is too much of a change for you. (I can always just use this forked code myself, as I don’t imagine there will be much changes in the future.)

Changes:
* Changes to ES6 modules loadable as-is by modern browsers; which means **no** IE support out-of-the-box. The reason for this is that _some_ form of exporting was necessary to consume these modules elsewhere. (CommonJS exports would always require some extra step/tooling to support browsers.)
* Adds some more JSDoc to allow the use of the TypeScript compiler to provide some type-checking and IDE features.
* Replaced minified version of `notie` with original ES6 version.
* Initial refactoring to move browser specific code out into its own module.

Considerations:
* Transpile to bundle that supports IE? Requires build-time tooling.
* Perhaps rename files as `.mjs` to signal they are ES6 modules?
* Perhaps change to use ES6 classes, which mostly means being able to have the TS compiler infer more with less docs needed (I _think_).
* Need to figure out the best way to provide input data, as the HTML5 `File` API does not exist natively in NodeJS (nor does any polyfill package seem to provide all the relied upon APIs).

Closes #41

### CODE REVIEW

File: js/CSVGood.js [hunk 1]
```diff
@@ -196,6 +196,9 @@ const CSVGood = function (file, onStep, onError, onComplete) {
 
         fileReader.onload = (evt) => {
             // Take result
+            if (!evt.target) {
+                throw new Error("Expected event to have a target")
+            }
```
Comment (alloy) [inline js/CSVGood.js:201]: I added all these types of checks to satisfy the type-checker errors thrown by the TS compiler where technically APIs can return `null`/`undefined`.

File: js/browser.js [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+                document.body.appendChild(link);
+
+                link.click();
+              }
+            }
+          });
+        }
+      });
+    }
+  };
```
Comment (alloy) [inline js/browser.js:57]: Basically a collection of all browser based code I collected from `convert.js`

**Comment by DanielsWrath** [comment]
Hi, Sorry for the late reply. Have been a bit busy.

Although I am very glad you put so much effort into this, I think these changes are too big for me personally. Especially since we are talking about a small feature. If I were to create an electron build of the system, I think I would be okay with this. However I believe the system is easier and just as safe to use in the browser browser.

You are of course free to fork this project and change all the code as needed.

The only changes that I can foresee (except bugfixes and added support for banks which are easily integrated into your fork) are a merge with the ABN AMRO version of YNABGoingDutch #25  and an integration into the YNAB api #26 . However these features are not guaranteed to be implemented ever.

I can always link to your repo in the readme.md if you'd like. Thanks again for the effort you put into this!

**Comment by alloy** [comment]
Makes total sense and no worries, I need this so there’s nothing lost here.

> The only changes that I can foresee (except bugfixes and added support for banks which are easily integrated into your fork) are a merge with the ABN AMRO version of YNABGoingDutch #25 and an integration into the YNAB api #26 . However these features are not guaranteed to be implemented ever.

I’m tracking the API one, but also don’t foresee it being a real problem to merge back in if it ever gets done.

> I can always link to your repo in the readme.md if you'd like.

Yeah that would be nice, I’m sure there will be some other dev at some point that wants this 👍 

> Thanks again for the effort you put into this!

Thank _you_ for the original work!

**Pull request closed without merging**